caikit/core/task.py

```diff
@@ -112,6 +104,47 @@ def get_output_type(cls) -> Type[DataBase]:
         """
         raise NotImplementedError("This is implemented by the @task decorator!")
 
+    @classmethod
+    def _raise_on_wrong_output_type(cls, output_type, module):
+        if cls._subclass_check(output_type, cls.get_output_type()):
+            # Basic case, same type or subclass of it
+            return
+
+        if typing.get_origin(output_type) == Union:
+            for union_type in typing.get_args(output_type):
+                if cls._subclass_check(union_type, cls.get_output_type()):
+                    # Something in the union has an acceptable type
+                    return
+
+        # Iterable output checks:
+        # `output_type` here can be any iterable, so long as it is type annotated with the type
+        # that is iterated over. So we first only check that it is iterable at all, and then look
+        # at the annotated type's args.
+        if isiterable(output_type) and isiterable(cls.get_output_type()):
+            # in this case we validated up-front that cls.output_type is Iterable[T]
+            streaming_type = typing.get_args(cls.get_output_type())[0]
+
+            for iterable_type in typing.get_args(output_type):
+                if cls._subclass_check(iterable_type, streaming_type):
+                    return
+
+        raise TypeError(
+            f"Wrong output type for module {module}: "
+            f"Found {output_type} but expected {cls.get_output_type()}"
+        )
+
+    @staticmethod
+    def _subclass_check(this_type, that_type):
+        """Wrapper around issubclass that first checks if both args are classes.
+        Returns True if the types are the same, or they are both classes and this_type
+        is a subclass of that_type
+        """
+        if this_type == that_type:
```

Comment:
unclear!

You're right that both this isn't a correct general-purpose implementation, and it should only be called on types from this package. I'm inclined to leave as is because we don't need a generally correct implementation?